File: sim/env_action.lua
-- Versioned learning-environment actions.
--
-- An action is the same abstract intent a human or a #112 bot expresses: a
-- continuous move vector plus named held actions and one-shot edges. It is
-- quantized into the canonical InputSample that sim.match already consumes, so a
-- policy cannot address resolver internals, teleport, set ownership or outcomes,
-- skip recovery or cooldowns, or act between fixed ticks. Anything a policy could
-- want that is not in this contract simply does not exist as an input.
--
-- Legality masks are derived from an observation, not from MatchState. A mask
-- built from a representative view therefore cannot contain privileged legality
-- by construction; a mask built from the privileged profile is tagged as such.

local input_frame = require("sim.input_frame")

---@alias EnvActionErrorCode
---| "malformed"
---| "move_out_of_range"
---| "unknown_held_action"
---| "unknown_edge_action"
---| "unavailable_action"

---@class EnvSlotAction
---@field version integer? -- Defaults to EnvAction.VERSION when omitted.
---@field move { x: number, y: number }? -- Continuous, inside the unit disc; nil means still.
---@field aim { x: number, y: number }? -- Continuous aim direction; nil means aim follows move.
---@field held table<InputHeldAction, boolean>? -- Continuously held intents for this tick.
---@field edges table<InputEdgeAction, boolean>? -- One-shot intents fired on this tick only.

---@class EnvActionMask
---@field version integer
---@field profile EnvObservationProfile
---@field privileged boolean -- True when the mask was derived from a privileged view.
---@field slot integer
---@field move boolean -- Movement is always legal; kept explicit for consumers.
---@field aim boolean -- Aiming is always legal; kept explicit for consumers.
---@field held table<InputHeldAction, boolean>
---@field edges table<InputEdgeAction, boolean>

---@class EnvActionModule
local env_action = {}

-- Bumped to 2 by #316: an action carries an independent aim direction, so a v1
-- action and a v2 action are not the same contract even when their fields agree.
env_action.VERSION = 2

---@type InputHeldAction[]
env_action.HELD_ACTIONS = {
    "shoot",
    "pass",
    "sprint",
    "jockey",
    "lob",
    "aerial_strike",
    "aerial_acrobatic",
    "equipment",
}

---@type InputEdgeAction[]
env_action.EDGE_ACTIONS = {
    "shoot",
    "pass",
    "switch",
    "dash",
    "dodge",
    "equipment_pressed",
    "equipment_released",
}

local ACTION_FIELDS = {
    version = true,
    move = true,
    aim = true,
    held = true,
    edges = true,
}

local MOVE_FIELDS = { x = true, y = true }
local AIM_FIELDS = { x = true, y = true }

-- Fixed-slot routing gives every slot exactly one player for the whole fixture,
-- so the legacy "switch to the outfielder nearest the ball" intent has no
-- meaning here. sim.match ignores it; the mask says so and the contract rejects
-- it with a reason instead of silently dropping it.
local UNAVAILABLE_EDGES = { switch = true }

local MOVE_EPSILON = 1e-9

---@param code EnvActionErrorCode
---@param message string
---@return nil, string, EnvActionErrorCode
local function reject(code, message)
    return nil, message, code
end

---@param value any
---@return boolean
local function is_finite(value)
    return type(value) == "number" and value == value and value < math.huge and value > -math.huge
end

---@param value any
---@param allowed table<string, boolean>
---@return boolean
local function only_known_fields(value, allowed)
    if type(value) ~= "table" then
        return false
    end
    for key in pairs(value) do
        if type(key) ~= "string" or not allowed[key] then
            return false
        end
    end
    return true
end

---@return EnvSlotAction
function env_action.neutral()
    return { version = env_action.VERSION, move = { x = 0, y = 0 }, held = {}, edges = {} }
end

-- Validate one slot action and return a normalized copy. Actions come from a
-- policy (external input), so an illegal action is a recoverable rejection with
-- a machine-readable reason, never an assert.
---@param action any
---@return EnvSlotAction?, string?, EnvActionErrorCode?
function env_action.validate(action)
    if not only_known_fields(action, ACTION_FIELDS) then
        return reject("malformed", "action must be a table of known fields")
    end
    if action.version ~= nil and action.version ~= env_action.VERSION then
        return reject("malformed", "unsupported env action version")
    end
    local move = { x = 0, y = 0 }
    if action.move ~= nil then
        if not only_known_fields(action.move, MOVE_FIELDS) then
            return reject("malformed", "action.move must be an { x, y } table")
        end
        local x = action.move.x or 0
        local y = action.move.y or 0
        if not is_finite(x) or not is_finite(y) then
            return reject("malformed", "action.move components must be finite numbers")
        end
        if x * x + y * y > 1 + MOVE_EPSILON then
            return reject("move_out_of_range", "action.move must lie inside the unit disc")
        end
        move = { x = x, y = y }
    end
    -- Aim is a direction, so unlike `move` it is not confined to the unit disc:
    -- only its angle survives quantization. A nil or zero-length aim is the
    -- absence of one, which `to_sample` encodes as AIM_NONE.
    local aim = nil
    if action.aim ~= nil then
        if not only_known_fields(action.aim, AIM_FIELDS) then
            return reject("malformed", "action.aim must be an { x, y } table")
        end
        local x = action.aim.x or 0
        local y = action.aim.y or 0
        if not is_finite(x) or not is_finite(y) then
            return reject("malformed", "action.aim components must be finite numbers")
        end
        if x ~= 0 or y ~= 0 then
            aim = { x = x, y = y }
        end
    end
    local held = {}
    if action.held ~= nil then
        if type(action.held) ~= "table" then
            return reject("malformed", "action.held must be a table")
        end
        for key, value in pairs(action.held) do
            if type(key) ~= "string" or input_frame.HELD_BITS[key] == nil then
                return reject("unknown_held_action", "unknown held action: " .. tostring(key))
            end
            if type(value) ~= "boolean" then
                return reject("malformed", "held action " .. key .. " must be a boolean")
            end
            held[key] = value
        end
    end
    local edges = {}
    if action.edges ~= nil then
        if type(action.edges) ~= "table" then
            return reject("malformed", "action.edges must be a table")
        end
        for key, value in pairs(action.edges) do
            if type(key) ~= "string" or input_frame.EDGE_BITS[key] == nil then
                return reject("unknown_edge_action", "unknown edge action: " .. tostring(key))
            end
            if type(value) ~= "boolean" then
                return reject("malformed", "edge action " .. key .. " must be a boolean")
            end
            edges[key] = value
        end
    end
    return { version = env_action.VERSION, move = move, aim = aim, held = held, edges = edges }
end

-- Drop the one-shot edges, keeping held intents. Used when one action is held
-- across several fixed ticks: an edge means "this tick only", so repeating it
-- would fabricate presses the policy never made.
---@param action EnvSlotAction
---@return EnvSlotAction
function env_action.without_edges(action)
    local held = {}
    for key, value in pairs(action.held or {}) do
        held[key] = value
    end
    local move = action.move or { x = 0, y = 0 }
    local aim = action.aim
    return {
        version = env_action.VERSION,
        move = { x = move.x, y = move.y },
        aim = aim and { x = aim.x, y = aim.y } or nil,
        held = held,
        edges = {},
    }
end

---@param action EnvSlotAction
---@return InputSample?, string?, EnvActionErrorCode?
function env_action.to_sample(action)
    local normalized, err, code = env_action.validate(action)
    if not normalized then
        return nil, err, code
    end
    local move = assert(normalized.move)
    local move_x, move_y, quantize_err = input_frame.quantize_move(move.x, move.y)
    if not move_x or not move_y then
        return reject("move_out_of_range", quantize_err or "action.move is not quantizable")
    end
    local held = 0
    local edges = 0
    for _, name in ipairs(env_action.HELD_ACTIONS) do
        if assert(normalized.held)[name] then
            held = held + input_frame.HELD_BITS[name]
        end
    end
    for _, name in ipairs(env_action.EDGE_ACTIONS) do
        if assert(normalized.edges)[name] then
            edges = edges + input_frame.EDGE_BITS[name]
        end
    end
    local aim = input_frame.AIM_NONE
    if normalized.aim ~= nil then
        local quantized, aim_err = input_frame.quantize_aim(normalized.aim.x, normalized.aim.y)
        if not quantized then
            return reject("malformed", aim_err or "action.aim is not quantizable")
        end
        aim = quantized
    end
    local sample, sample_err = input_frame.new_sample({
        move_x = move_x,
        move_y = move_y,
        held = held,
        edges = edges,
        aim = aim,
    })
    if not sample then
        return reject("malformed", sample_err or "action is not encodable as an InputSample")
    end
    return sample
end

---@param sample InputSample
---@return EnvSlotAction?, string?, EnvActionErrorCode?
function env_action.from_sample(sample)
    local move_x, move_y, move_err = input_frame.dequantize_move(sample)
    if not move_x or not move_y then
        return reject("malformed", move_err or "sample move is not decodable")
    end
    local held = {}
    for _, name in ipairs(env_action.HELD_ACTIONS) do
        held[name] = input_frame.is_held(sample, name) == true
    end
    local edges = {}
    for _, name in ipairs(env_action.EDGE_ACTIONS) do
        edges[name] = input_frame.has_edge(sample, name) == true
    end
    local aim = nil
    if sample.aim ~= input_frame.AIM_NONE then
        local aim_x, aim_y, aim_err = input_frame.dequantize_aim(sample.aim)
        if not aim_x or not aim_y then
            return reject("malformed", aim_err or "sample aim is not decodable")
        end
        aim = { x = aim_x, y = aim_y }
    end
    return {
        version = env_action.VERSION,
        move = { x = move_x, y = move_y },
        aim = aim,
        held = held,
        edges = edges,
    }
end

-- Legality a normal client can know, computed only from the fields present in
-- `view`. The representative and team profiles carry no private timers, so the
-- resulting mask cannot encode privileged legality; the privileged profile is
-- flagged so a report can never present it as a human-equivalent mask.
--
-- Only `own` and `ball` are read, so a narrow EnvActionView is enough: callers on
-- the per-step path should pass one instead of building a whole observation.
---@param view EnvSlotView|EnvActionView
---@return EnvActionMask?, string?, EnvActionErrorCode?
function env_action.mask(view)
    if type(view) ~= "table" or type(view.own) ~= "table" or type(view.ball) ~= "table" then
        return reject("malformed", "an action mask needs a slot view")
    end
    local own = view.own
    local equipment = own.equipment
    local equipped = equipment ~= nil
    local equipment_ready = equipment ~= nil and equipment.ready == true
    local aerial_available = own.header_ready and view.ball.airborne and not own.stunned
    return {
        version = env_action.VERSION,
        profile = view.profile,
        privileged = view.profile == "privileged",
        slot = view.slot,
        move = true,
        aim = true,
        held = {
            shoot = true,
            pass = true,
            sprint = true,
            jockey = true,
            lob = true,
            aerial_strike = aerial_available,
            aerial_acrobatic = aerial_available,
            equipment = equipped,
        },
        edges = {
            shoot = true,
            pass = true,
            switch = false,
            dash = own.tackle_ready and not own.stunned,
            dodge = own.dodge_ready and not own.stunned,
            equipment_pressed = equipment_ready,
            equipment_released = equipped,
        },
    }
end

-- Check a validated action against a mask and name the first violation. The
-- environment reports this reason back to the caller so a policy learns why an
-- intent was refused instead of silently losing it.
---@param action EnvSlotAction
---@param mask EnvActionMask
---@return boolean?, string?, EnvActionErrorCode?
function env_action.check_mask(action, mask)
    assert(type(mask) == "table", "an action mask is required")
    local normalized, err, code = env_action.validate(action)
    if not normalized then
        return nil, err, code
    end
    for _, name in ipairs(env_action.HELD_ACTIONS) do
        if assert(normalized.held)[name] and not mask.held[name] then
            return reject(
                "unavailable_action",
                "held action " .. name .. " is not available to slot " .. mask.slot .. " this tick"
            )
        end
    end
    for _, name in ipairs(env_action.EDGE_ACTIONS) do
        if assert(normalized.edges)[name] and not mask.edges[name] then
            local reason = UNAVAILABLE_EDGES[name]
                    and ("edge action " .. name .. " does not exist under fixed-slot routing; slot " .. mask.slot .. " controls one player for the whole fixture")
                or (
                    "edge action "
                    .. name
                    .. " is not available to slot "
                    .. mask.slot
                    .. " this tick"
                )
            return reject("unavailable_action", reason)
        end
    end
    return true
end

return env_action

File: rust/crates/gc-sim/tests/env_action.rs
//! Port of `spec/sim/env_action_spec.lua`.
//!
//! The Lua spec builds its `mask`-family fixtures via `sim.env.reset` /
//! `sim.env.observe`, which need `sim.match` and `sim.combat` — neither
//! ported yet (v2/README §1, task scope for this module explicitly excludes
//! `env.lua`/`env_observation.lua`). `env_action.mask`/`check_mask` only ever
//! read `view.own`/`view.ball`, so every assertion below is preserved by
//! constructing an [`EnvActionView`] with exactly the fields the Lua fixture
//! would have produced, instead of driving a match to produce one. This is
//! on the wire path (`to_sample`/`from_sample` feed the same `InputSample`
//! that crosses rollback resim); see `differential.rs` for the required
//! bit-for-bit comparison against the reference Lua implementation for the
//! `input_frame` primitives this module builds on.

use gc_sim::env_action::{
    self, EnvActionErrorCode, EnvActionView, EnvActionViewBall, EnvActionViewEquipment,
    EnvActionViewSelf, EnvObservationProfile, RawAction, RawTable, RawValue,
};
use gc_sim::input_frame;

fn table(pairs: Vec<(&str, RawValue)>) -> RawTable {
    pairs
        .into_iter()
        .map(|(key, value)| (key.to_string(), value))
        .collect()
}

fn num(value: f64) -> RawValue {
    RawValue::Number(value)
}

fn boolean(value: bool) -> RawValue {
    RawValue::Bool(value)
}

fn nested(pairs: Vec<(&str, RawValue)>) -> RawValue {
    RawValue::Table(table(pairs))
}

#[test]
fn env_action_validate_normalizes_a_sparse_action_into_the_full_contract() {
    let raw = RawAction::Table(table(vec![
        ("move", nested(vec![("x", num(0.5))])),
        ("held", nested(vec![("sprint", boolean(true))])),
    ]));
    let action = env_action::validate(&raw).unwrap();
    assert_eq!(action.version, env_action::VERSION);
    assert_eq!(action.movement.x, 0.5);
    assert_eq!(action.movement.y, 0.0);
    assert!(action.held.sprint);
    assert!(!action.held.lob);
    assert_eq!(action.edges, Default::default());
}

#[test]
fn env_action_validate_treats_a_neutral_action_as_a_real_no_op() {
    let sample = env_action::to_sample(&env_action::neutral()).unwrap();
    let neutral = input_frame::neutral_sample();
    assert_eq!(sample.move_x, neutral.move_x);
    assert_eq!(sample.move_y, neutral.move_y);
    assert_eq!(sample.held, neutral.held);
    assert_eq!(sample.edges, neutral.edges);
}

#[test]
fn env_action_validate_rejects_malformed_actions_with_machine_readable_reasons() {
    let cases: Vec<(RawAction, EnvActionErrorCode)> = vec![
        (RawAction::Other, EnvActionErrorCode::Malformed),
        (
            RawAction::Table(table(vec![("throttle", num(1.0))])),
            EnvActionErrorCode::Malformed,
        ),
        (
            RawAction::Table(table(vec![("version", num(99.0))])),
            EnvActionErrorCode::Malformed,
        ),
        (
            RawAction::Table(table(vec![(
                "move",
                nested(vec![("x", num(0.0)), ("y", num(0.0)), ("z", num(1.0))]),
            )])),
            EnvActionErrorCode::Malformed,
        ),
        (
            RawAction::Table(table(vec![(
                "move",
                nested(vec![("x", num(f64::INFINITY))]),
            )])),
            EnvActionErrorCode::Malformed,
        ),
        (
            RawAction::Table(table(vec![(
                "move",
                nested(vec![("x", num(1.0)), ("y", num(1.0))]),
            )])),
            EnvActionErrorCode::MoveOutOfRange,
        ),
        (
            RawAction::Table(table(vec![(
                "held",
                nested(vec![("teleport", boolean(true))]),
            )])),
            EnvActionErrorCode::UnknownHeldAction,
        ),
        (
            RawAction::Table(table(vec![("held", nested(vec![("sprint", num(1.0))]))])),
            EnvActionErrorCode::Malformed,
        ),
        (
            RawAction::Table(table(vec![(
                "edges",
                nested(vec![("set_owner", boolean(true))]),
            )])),
            EnvActionErrorCode::UnknownEdgeAction,
        ),
        (
            RawAction::Table(table(vec![(
                "edges",
                nested(vec![("dash", RawValue::Table(table(vec![])))]),
            )])),
            EnvActionErrorCode::Malformed,
        ),
    ];
    for (raw, expected_code) in cases {
        let err = env_action::validate(&raw).expect_err("the action must be rejected");
        assert_eq!(err.code, expected_code);
        assert!(!err.message.is_empty(), "a reason is always supplied");
    }
}

#[test]
fn env_action_validate_accepts_the_unit_disc_boundary() {
    assert!(
        env_action::validate(&RawAction::Table(table(vec![(
            "move",
            nested(vec![("x", num(1.0)), ("y", num(0.0))]),
        )])))
        .is_ok()
    );
    assert!(
        env_action::validate(&RawAction::Table(table(vec![(
            "move",
            nested(vec![("x", num(0.6)), ("y", num(0.8))]),
        )])))
        .is_ok()
    );
}

#[test]
fn env_action_sample_conversion_round_trips_through_the_canonical_quantization() {
    let raw = RawAction::Table(table(vec![
        ("move", nested(vec![("x", num(1.0)), ("y", num(0.0))])),
        (
            "held",
            nested(vec![
                ("sprint", boolean(true)),
                ("lob", boolean(true)),
                ("equipment", boolean(true)),
            ]),
        ),
        (
            "edges",
            nested(vec![
                ("dash", boolean(true)),
                ("equipment_pressed", boolean(true)),
            ]),
        ),
    ]));
    let action = env_action::validate(&raw).unwrap();
    let sample = env_action::to_sample(&action).unwrap();
    assert!(input_frame::is_held(&sample, input_frame::HeldAction::Sprint).unwrap());
    assert!(input_frame::is_held(&sample, input_frame::HeldAction::Lob).unwrap());
    assert!(input_frame::is_held(&sample, input_frame::HeldAction::Equipment).unwrap());
    assert!(!input_frame::is_held(&sample, input_frame::HeldAction::Jockey).unwrap());
    assert!(input_frame::has_edge(&sample, input_frame::EdgeAction::Dash).unwrap());
    assert!(input_frame::has_edge(&sample, input_frame::EdgeAction::EquipmentPressed).unwrap());
    assert!(!input_frame::has_edge(&sample, input_frame::EdgeAction::Switch).unwrap());

    let decoded = env_action::from_sample(&sample).unwrap();
    assert!(decoded.held.sprint);
    assert!(!decoded.held.jockey);
    assert!(decoded.edges.dash);
    let re_sample = env_action::to_sample(&decoded).unwrap();
    assert_eq!(re_sample.held, sample.held);
    assert_eq!(re_sample.move_x, sample.move_x);
}

#[test]
fn env_action_sample_conversion_drops_one_shot_edges_when_an_action_is_held_across_ticks() {
    let raw = RawAction::Table(table(vec![
        ("move", nested(vec![("x", num(-1.0)), ("y", num(0.0))])),
        ("held", nested(vec![("jockey", boolean(true))])),
        ("edges", nested(vec![("dodge", boolean(true))])),
    ]));
    let action = env_action::validate(&raw).unwrap();
    let held = env_action::without_edges(&action);
    assert!(held.held.jockey);
    assert_eq!(held.edges, Default::default());
    assert_eq!(held.movement.x, -1.0);
}

/// A fresh, unconstrained slot: nothing on cooldown, nothing equipped, ball
/// grounded. Mirrors the fixture the Lua spec gets from `env.reset` before
/// any mutation.
fn fresh_view() -> EnvActionView {
    EnvActionView {
        profile: EnvObservationProfile::Representative,
        slot: 1,
        own: EnvActionViewSelf {
            stunned: false,
            header_ready: true,
            tackle_ready: true,
            dodge_ready: true,
            equipment: None,
        },
        ball: EnvActionViewBall { airborne: false },
    }
}

#[test]
fn env_action_mask_derives_legality_from_the_view_alone() {
    let mask = env_action::mask(&fresh_view());
    assert_eq!(mask.version, env_action::VERSION);
    assert_eq!(mask.slot, 1);
    assert!(mask.movement);
    assert!(!mask.privileged);
    assert!(mask.held.shoot);
    assert!(mask.held.pass);
    assert!(mask.held.sprint);
    assert!(mask.edges.dash, "a fresh player may challenge");
    assert!(mask.edges.dodge);
    assert!(
        !mask.edges.switch,
        "fixed-slot routing has no player switch"
    );
    assert!(!mask.held.equipment, "no loadout, no equipment intent");
    assert!(!mask.edges.equipment_pressed);
}

#[test]
fn env_action_mask_closes_gated_intents_while_the_observable_cooldown_runs() {
    let mut view = fresh_view();
    view.own.tackle_ready = false;
    view.own.dodge_ready = false;
    let gated = env_action::mask(&view);
    assert!(!gated.edges.dash);
    assert!(!gated.edges.dodge);

    view.own.tackle_ready = true;
    view.own.dodge_ready = true;
    view.own.stunned = true;
    let stunned = env_action::mask(&view);
    assert!(!stunned.edges.dash);
    assert!(!stunned.edges.dodge);
}

#[test]
fn env_action_mask_opens_aerial_intents_only_while_the_ball_is_airborne() {
    let mut view = fresh_view();
    assert!(!env_action::mask(&view).held.aerial_strike);
    view.ball.airborne = true;
    let airborne = env_action::mask(&view);
    assert!(airborne.held.aerial_strike);
    assert!(airborne.held.aerial_acrobatic);
}

#[test]
fn env_action_mask_gates_equipment_on_the_observable_combat_readiness() {
    let mut view = fresh_view();
    view.own.equipment = Some(EnvActionViewEquipment { ready: true });
    let ready = env_action::mask(&view);
    assert!(ready.held.equipment);
    assert!(ready.edges.equipment_pressed);

    view.own.equipment = Some(EnvActionViewEquipment { ready: false });
    let cooling = env_action::mask(&view);
    assert!(!cooling.edges.equipment_pressed);
    assert!(
        cooling.held.equipment,
        "the equipment still exists, it is just not ready"
    );
}

#[test]
fn env_action_mask_rejects_a_masked_out_intent_with_a_reason_naming_the_slot() {
    let mask = env_action::mask(&fresh_view());

    let switch = RawAction::Table(table(vec![(
        "edges",
        nested(vec![("switch", boolean(true))]),
    )]));
    let switch_action = env_action::validate(&switch).unwrap();
    let err = env_action::check_mask(&switch_action, &mask).expect_err("switch is never legal");
    assert_eq!(err.code, EnvActionErrorCode::UnavailableAction);
    assert!(err.message.contains("fixed-slot routing"));

    let allowed_raw = RawAction::Table(table(vec![
        ("move", nested(vec![("x", num(1.0))])),
        ("edges", nested(vec![("dash", boolean(true))])),
    ]));
    let allowed_action = env_action::validate(&allowed_raw).unwrap();
    assert!(env_action::check_mask(&allowed_action, &mask).unwrap());

    let aerial_raw = RawAction::Table(table(vec![(
        "held",
        nested(vec![("aerial_strike", boolean(true))]),
    )]));
    let aerial_action = env_action::validate(&aerial_raw).unwrap();
    let aerial_err =
        env_action::check_mask(&aerial_action, &mask).expect_err("aerial is not open here");
    assert_eq!(aerial_err.code, EnvActionErrorCode::UnavailableAction);
}
